src/Compilers/CSharp/Portable/Compilation/CSharpDiagnosticFilter.cs

```diff
@@ -117,6 +117,63 @@ internal static class CSharpDiagnosticFilter
             return d.WithReportDiagnostic(reportAction);
         }
 
+        internal static ReportDiagnostic FilterAndGetReportDiagnostic(
```

Comment:
I'm wondering if there's a better way to factor the code with the `Filter` method this is modeled after. It feels like we could reduce duplication.
For example, could we extract a shared helper from `Filter` to take a `DiagnosticDescriptor` and a `Location`? The logic here would call that helper (with the description and `Location.None`) then either return `ReportDiagnostics.Suppress` or result from calling `MapSeverityToReport`.

Reply:
I tried something like that originally and ended up getting a bit lost. Let me shake off the cobwebs and give it another go.